src/modules/flow-metatype/http-composed-server/http-composed-server.c

```diff
@@ -1050,58 +1086,246 @@ http_composed_server_ports_description(const struct sol_flow_metatype_context *c
     return 0;
+}
 
-err_port_data:
-    free(opt_port->name);
-err_port_name:
-    sol_vector_del_last(options);
-err_port:
-    free(opt_path->data_type);
-err_data:
-    free(opt_path->name);
-err:
-    sol_vector_del_last(options);
-    return -ENOMEM;
+static int
+http_composed_server_generate_end(const struct sol_flow_metatype_context *ctx,
+    struct sol_buffer *out)
+{
+    return 0;
+}
+
+static int
+setup_get_port_function(struct sol_buffer *out, struct sol_vector *ports,
+    const struct sol_str_slice prefix, const char *port_type)
+{
+    int r;
+    uint16_t i;
+    struct sol_flow_metatype_port_description *port;
+
+    r = sol_buffer_append_printf(out,
+        "static const struct sol_flow_port_type_%s *\n"
+        "http_composed_server_%.*s_get_%s_port(const struct sol_flow_node_type *type, uint16_t port)\n"
+        "{\n", port_type, SOL_STR_SLICE_PRINT(prefix), port_type);
+    SOL_INT_CHECK(r, < 0, r);
+
+    SOL_VECTOR_FOREACH_IDX (ports, port, i) {
+        r = sol_buffer_append_printf(out, "    if (port == %u)\n"
+            "        return &http_composed_server_%.*s_%s_port.base;\n",
+            i, SOL_STR_SLICE_PRINT(prefix), port->name);
+        SOL_INT_CHECK(r, < 0, r);
+    }
+
+    return sol_buffer_append_slice(out, sol_str_slice_from_str("    return NULL;\n}\n"));
+}
+
+static int
+setup_composed_packet(struct sol_buffer *out, const struct sol_str_slice prefix,
+    const struct sol_str_slice types, const char *port_name)
+{
+    int r;
+    struct sol_vector tokens;
+    struct sol_str_slice *token;
+    uint16_t i;
+
+    r = sol_buffer_append_slice(out,
+        sol_str_slice_from_str("        const struct sol_flow_packet_type *types[] = {"));
+    SOL_INT_CHECK(r, < 0, r);
+
+    tokens = sol_str_slice_split(types, ",", 0);
+
+    SOL_VECTOR_FOREACH_IDX (&tokens, token, i) {
+        r = sol_buffer_append_printf(out, "%s,",
+            sol_flow_packet_get_packet_type_as_string(*token));
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
+    }
+
+    r = sol_buffer_append_printf(out, "NULL};\n"
+        "        http_composed_server_%.*s_%s_port.base.packet_type = sol_flow_packet_type_composed_new(types);\n",
+        SOL_STR_SLICE_PRINT(prefix), port_name);
+
+exit:
+    sol_vector_clear(&tokens);
+    return r;
+}
+
+static int
+setup_packet_type(struct sol_buffer *out, struct sol_vector *ports,
+    const struct sol_str_slice prefix)
+{
+    int r;
+    uint16_t i;
+    struct sol_flow_metatype_port_description *port;
+
+    SOL_VECTOR_FOREACH_IDX (ports, port, i) {
+        r = sol_buffer_append_printf(out,
+            "    if (!http_composed_server_%.*s_%s_port.base.packet_type) {\n",
+            SOL_STR_SLICE_PRINT(prefix), port->name);
+        SOL_INT_CHECK(r, < 0, r);
+
+        if (strstartswith(port->type, "composed:")) {
+            struct sol_str_slice types;
+            //Removing the composed: prefix
+            types.data = port->type + 9;
+            types.len = strlen(port->type) - 9;
```

Comment:
It's different, thi code is for the generator, the function that was added returns an object.